File: EmployeeImportance690.java
//Time Complexity: O(N) 
//Space Complexity: O(N) + O(N) (hashmap and queue)
//Leetcode: Yes
/*
// Definition for Employee.
class Employee {
    public int id;
    public int importance;
    public List<Integer> subordinates;
};
*/

class Solution {
    public int getImportance(List<Employee> employees, int id) {
        int total_importance = 0;
        HashMap<Integer, Employee> map = new HashMap<Integer, Employee>();
        //create a maping for employee id  to entire empyoee data.  id 1 ---> employee 1 data
        for (Employee emp : employees) {
            map.put(emp.id, emp);
        }
        
        Queue<Employee> queue = new LinkedList<Employee>();
        //put the given empoyee in the queue
        queue.add(map.get(id));
        total_importance = map.get(id).importance;
        
        while (!queue.isEmpty()) {
            Employee e = queue.poll();
            for (Integer sub : e.subordinates) {
                total_importance += map.get(sub).importance;
                queue.add(map.get(sub));    
            }
        }
        return total_importance;
    }
}

File: employee-importance_690.py
#Approach: BFS
#Mapping employee object to its ID for O(1) search, appending object and adding their and its subordinate's importance
#Time Complexity O(V+E) where V is nodes of Employees and E is connections to subordinates
#Space Complexity O(V+E)
#It successfully runs on leetcode 

from collections import deque
class Solution:
    def getImportance(self, employees: List['Employee'], id: int) -> int:
        if not employees: return 0
        l={}
        q= deque()
        for i in employees:
            l[i.id]= i
        q.append(l[id])
        imp=0
        while q:
            for i in range(len(q)):
                cur = q.popleft()
                imp+=cur.importance    
                for i in cur.subordinates:
                    q.append(l[i])
        return imp

#Approach: DFS
#Mapping employee object to its ID for O(1) search,soing dfs on each object and its subordinate's
#Time Complexity O(V+E) where V is nodes of Employees and E is connections to subordinates
#Space Complexity O(V+E)
#It successfully runs on leetcode 

from collections import deque
class Solution:
    def getImportance(self, employees: List['Employee'], id: int) -> int:
        if not employees: return 0
        l={}
        for i in employees:
            l[i.id]= i
        if not l.get(id): return 0 
        self.result = l[id].importance
        self.dfs(l,l[id])
        return self.result
        
    def dfs(self,l, e):
        if not e.subordinates: return self.result
        for i in e.subordinates:
            self.result+=l[i].importance
            self.dfs(l,l[i])